src/reactionsystem_conversions.jl

```diff
@@ -0,0 +1,900 @@
+            push!(meqs, makemajump(rx; combinatoric_ratelaw = combinatoric_ratelaws))
+        else
+            rl = jumpratelaw(rx; combinatoric_ratelaw = combinatoric_ratelaws)
+            affect = Vector{Equation}()
+            for (spec, stoich) in rx.netstoich
+                # don't change species that are constant or BCs
+                (!drop_dynamics(spec)) && push!(affect, spec ~ spec + stoich)
+            end
+            if isvrj
+                push!(veqs, VariableRateJump(rl, affect))
+            else
+                push!(ceqs, ConstantRateJump(rl, affect))
+            end
+        end
+    end
+    vcat(meqs, ceqs, veqs)
+end
+
+
+### Equation Coupling ###
+
+# merge constraint components with the ReactionSystem components
+# also handles removing BC and constant species
+function addconstraints!(eqs, rs::ReactionSystem, ists, ispcs; remove_conserved = false)
+    # if there are BC species, put them after the independent species
+    rssts = get_unknowns(rs)
+    sts = any(isbc, rssts) ? vcat(ists, filter(isbc, rssts)) : ists
+    ps = get_ps(rs)
+
+    # make dependent species observables and add conservation constants as parameters
+    if remove_conserved
+        nps = get_networkproperties(rs)
+
+        # add the conservation constants as parameters and set their values
+        ps = vcat(ps, collect(eq.lhs for eq in nps.constantdefs))
+        defs = copy(MT.defaults(rs))
+        for eq in nps.constantdefs
+            defs[eq.lhs] = eq.rhs
+        end
+
+        # add the dependent species as observed
+        obs = copy(MT.observed(rs))
+        append!(obs, nps.conservedeqs)
+    else
+        defs = MT.defaults(rs)
+        obs = MT.observed(rs)
+    end
+
+    ceqs = Equation[eq for eq in get_eqs(rs) if eq isa Equation]
+    if !isempty(ceqs)
+        if remove_conserved
+            @info """
+                  Be careful mixing constraints and elimination of conservation laws.
+                  Catalyst does not check that the conserved equations still hold for the
+                  final coupled system of equations. Consider using `remove_conserved =
+                  false` and instead calling ModelingToolkit.structural_simplify to simplify
+                  any generated ODESystem or NonlinearSystem.
+                  """
+        end
+        append!(eqs, ceqs)
+    end
+
+    eqs, sts, ps, obs, defs
+end
+
+# used by flattened systems that don't support constraint equations currently
+function error_if_constraints(::Type{T}, sys::ReactionSystem) where {T <: MT.AbstractSystem}
+    any(eq -> eq isa Equation, get_eqs(sys)) &&
+        error("Can not convert to a system of type ", T,
+              " when there are constraint equations.")
+    nothing
+end
+
+
+### Utility ###
+
+# Function returning `true` for species which shouldn't change from the reactions, 
+# including non-species variables.
+drop_dynamics(s) = isconstant(s) || isbc(s) || (!isspecies(s))
```

Comment:
Please place before `assemble_oderhs` where it is first used.

Reply:
can do